src/unix/lwt_unix_unix.h

```diff
@@ -3009,38 +3014,200 @@ CAMLprim value lwt_unix_bind_job(value fd, value address)
 static void worker_getcwd(struct job_getcwd *job)
 {
+    /* Run the C call. We don't perform any checking in the worker thread,
+       because we typically don't want to take any other action here – we want
+       to take action in the main thread. In more complex calls, pre-checks on
+       the arguments are done in the `lwt_unix_FOO_job` job-allocating
+       function, and post-checks on the results are done in the `result_FOO`
+       function. */
     job->result = getcwd(job->buf, sizeof(job->buf));
+
+    /* Store the current value of `errno`. Note that if the C call succeeded, it
+       did not reset `errno` to zero. In that case, `errno` still contains the
+       error code from the last C call to fail in this worker thread. This means
+       that `errno`/`job->error_code` *cannot* be used to determine whether the
+       C call succeeded or not. */
     job->error_code = errno;
 }
 
+/* Runs in the main thread. This function is `static` for the same reason as
+   `worker_getcwd`. */
 static value result_getcwd(struct job_getcwd *job)
 {
+    /* This macro is defined in `lwt_unix.h`. The arguments are used as follows:
+
+       - The first argument is the name of the job variable.
+       - If the check in the second argument *succeds*, the C call, and job,
+         failed (confusing!). Note that this check must *not* be based solely on
+         `job->error_code`; see comment in `worker_getcwd` above.
+       - The last argument is the name of the C call, used in a
+         `Unix.Unix_error` exception raised if the job failed.
+
+       If the check succeeds/job failed, this macro deallocates the job, raises
+       the exception, and does *not* "return" to the rest of `result_getcwd`.
+       Otherwise, if the job succeeded, the job is *not* deallocated, and
+       execution continues in the rest of `result_getcwd`.
+
+       `job->error_code` is used internally by the macro in creating the
+       `Unix.Unix_error`. If this is incorrect (i.e., some job does not set
+       `errno` on failure), it is necessary to replace the macro by its
+       expansion, and modify the behavior. */
     LWT_UNIX_CHECK_JOB(job, job->result == NULL, "getcwd");
+
+    /* Create an OCaml string from the temporary buffer into which the
+       `getcwd(3)` wrote the current directory. This copies the data.
+
+       Throughout Lwt, blocking C calls run in worker threads can't write
+       directly into OCaml strings, because the OCaml garbage collector might
+       move the strings after the pointer has already been passed to the call,
+       but while the call is still blocked. Bigarrays don't have this problem,
+       so pointers into them are passed to blocking C calls, saving a copy.
+
+       For jobs that return integers or other kinds of values, it is necessary
+       to use the various `Int_val`, `Long_val` macros, etc. See
+
+         http://caml.inria.fr/pub/docs/manual-ocaml/intfc.html#sec415 */
```

Comment:
Maybe start with a generic recommendation (e.g., “Convert the result of the job to its OCaml representation”) and only after  that talk about the specific of this particular job (strings) and other values (bigarrays, ints, etc.).